File: src/main/java/net/lifecity/mc/skillmaster/skill/separatedskills/wall/Wall.java
package net.lifecity.mc.skillmaster.skill.separatedskills.wall;

import com.sk89q.worldedit.extent.clipboard.Clipboard;
import net.lifecity.mc.skillmaster.SkillMaster;
import net.lifecity.mc.skillmaster.skill.SeparatedSkill;
import net.lifecity.mc.skillmaster.skill.SkillType;
import net.lifecity.mc.skillmaster.user.SkillUser;
import net.lifecity.mc.skillmaster.utils.WorldEditUtils;
import net.lifecity.mc.skillmaster.weapon.Weapon;
import org.bukkit.Location;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.Player;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;

public class Wall extends SeparatedSkill {
    public Wall(Weapon weapon,
                   int num,
                   int point,
                   int activationTime,
                   int interval,
                   SkillUser user
    ) {
        super(
                "ウォール",
                weapon,
                SkillType.ATTACK,
                Arrays.asList(
                        "ガラスの球体を生成し、相手の攻撃から身を守る",
                        "球体は時間経過で消える"
                ),
                num,
                point,
                activationTime,
                interval,
                user
        );
    }

    @Override
    public void activate() {
        super.activate();

        Player player = this.user.getPlayer();
        Location loc = player.getLocation();
        try {
            createWall(loc);
        } catch (IOException e) {
            player.sendMessage("壁の生成失敗");
        }
    }

    @Override
    public void additionalInput() {}

    @Override
    public void deactivate() {
        super.deactivate();
    }

    private void createWall(Location origin) throws IOException {
        float yaw = origin.getYaw();
        BlockFace face = yawToFace(yaw, true);
        WorldEditUtils worldEditUtils = new WorldEditUtils();


        switch(face) {
            case NORTH -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/north_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case NORTH_EAST -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/north_east_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case EAST -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/east_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case SOUTH_EAST -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/south_east_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case SOUTH -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/south_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case SOUTH_WEST -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/south_west_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case WEST -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/west_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
            case NORTH_WEST -> {
                File schematic = new File(SkillMaster.instance.getDataFolder() + File.separator + "/schematics/north_west_wall.schem");
                Clipboard clip = worldEditUtils.load(schematic);
                worldEditUtils.paste(origin, clip);
            }
        }
    }


    private BlockFace yawToFace(float yaw, boolean useSubCardinalDirections) {
        if (useSubCardinalDirections)
            return radial[Math.round(yaw / 45f) & 0x7].getOppositeFace();

        return axis[Math.round(yaw / 90f) & 0x3].getOppositeFace();
    }

    private static final BlockFace[] axis = { BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST };
    private static final BlockFace[] radial = { BlockFace.NORTH, BlockFace.NORTH_EAST, BlockFace.EAST, BlockFace.SOUTH_EAST, BlockFace.SOUTH, BlockFace.SOUTH_WEST, BlockFace.WEST, BlockFace.NORTH_WEST };
}

File: src/main/java/net/lifecity/mc/skillmaster/skill/compositeskills/Wall.kt
package net.lifecity.mc.skillmaster.skill.compositeskills

import net.lifecity.mc.skillmaster.SkillMaster
import net.lifecity.mc.skillmaster.skill.ICompositeSkill
import net.lifecity.mc.skillmaster.skill.SkillManager
import net.lifecity.mc.skillmaster.skill.SkillType
import net.lifecity.mc.skillmaster.user.SkillUser
import net.lifecity.mc.skillmaster.utils.WorldEditUtils
import net.lifecity.mc.skillmaster.weapon.Weapon
import org.bukkit.Location
import org.bukkit.block.BlockFace
import org.bukkit.block.BlockFace.*
import java.io.File
import java.io.IOException
import kotlin.math.roundToInt

class Wall(
    override val activationTime: Int = 10,
    override val canCancel: Boolean = true,
    override val name: String = "ウォール",
    override val weaponList: List<Weapon> = listOf(
        Weapon.STRAIGHT_SWORD,
        Weapon.DAGGER,
        Weapon.GREAT_SWORD,
        Weapon.LONG_SWORD,
        Weapon.RAPIER,
        Weapon.MACE
    ),
    override val type: SkillType = SkillType.ATTACK,
    override val lore: List<String> = listOf(
        "ガラスの壁を生成し、相手の攻撃から身を守る",
        "壁は時間経過で消える"
    ),
    override var isActivated: Boolean = false,
    override val interval: Int = 240,
    override var inInterval: Boolean,
    override val user: SkillUser
) : ICompositeSkill {
    override fun onAdditionalInput() {
    }

    override fun register() {
        SkillManager(this).register()
    }

    override fun canActivate(): Boolean  = true

    override fun onActivate() {
        try {
            createWall(user.player.location)
        } catch (e: IOException) {
            user.player.sendMessage("壁の生成失敗")
        }
    }

    override fun onDeactivate() {
    }


    @Throws(IOException::class)
    private fun createWall(origin: Location) {
        val yaw = origin.yaw
        val face = yawToFace(yaw)
        val worldEditUtils = WorldEditUtils()
        when (face) {
            NORTH -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/north_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            NORTH_EAST -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/north_east_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            EAST -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/east_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            SOUTH_EAST -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/south_east_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            SOUTH -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/south_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            SOUTH_WEST -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/south_west_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            WEST -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/west_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            NORTH_WEST -> {
                val schematic =
                    File(SkillMaster.INSTANCE.dataFolder.toString() + File.separator + "/schematics/north_west_wall.schem")
                val clip = worldEditUtils.load(schematic)
                worldEditUtils.pasteAndAutoUndo(origin, clip, activationTime)
            }

            else -> {}
        }
    }

    private fun yawToFace(yaw: Float): BlockFace {
        return radial[(yaw / 45f).roundToInt() and 0x7].oppositeFace
    }

    companion object {
        private val radial = arrayOf(
            NORTH,
            NORTH_EAST,
            EAST,
            SOUTH_EAST,
            SOUTH,
            SOUTH_WEST,
            WEST,
            NORTH_WEST
        )
    }
}